docs/enhancements/cluster-install-apis.md

```diff
@@ -0,0 +1,235 @@
+Failed | True once an install attempt has failed. Does not imply we've stopped trying, if retries are supported by the controller we may be re-trying. Returns to False if we later reach Completed=True.
+Stopped | True once the controller is no longer working to reach desired state. Does not imply success or failure by itself, but can be combined with Completed/Failed.
+
+These interface conditions will be copied back onto ClusterDeployment by Hive controllers with a "ClusterInstall" prefix, the ClusterInstall controller does not need to worry about this and in general should never write to ClusterDeployment. example: Failed becomes ClusterInstallFailed (which exists today on ClusterDeployment for this purpose).
+
+In the case of ClusterInstallFailed, we will also copy this condition to ProvisionFailed to maintain the current contract and not affect callers monitoring for this condition.
+
+ClusterInstall controllers can add their own conditions, but these will not transfer back to the ClusterDeployment. UIs around this process may need to show ClusterInstall conditions explicitly.
+
+The core ClusterInstall CRDs and controllers discussed in this document should live in Hive. However, in theory, it would be possible for an external application to implement it's own ClusterInstall outside of Hive and still interface with ClusterDeployment using this system.
+
+#### External ClusterInstall RBAC
+
+To support an external application providing their own ClusterInstall CRD implementation, Hive will need RBAC to be able to access those objects. This can be achieved by using [Aggregated ClusterRoles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#aggregated-clusterroles).
+
+Hive will be updated to deploy an empty ClusterRole `hive-controllers-clusterinstall` which will contain no rules, but will match the label `hive.openshift.io/aggregate-to-hive-clusterinstall`. Integrating components will need to deploy a ClusterRole with this label and rules that provide Hive all permissions on their custom ClusterInstall CRD.
+
+This will not be required for ClusterInstall implementations that live within Hive. (including AgentClusterInstall)
+
+### Cluster Deprovisioning
+
+Today in Hive each ClusterDeployment is given a finalizer automatically, and when the API resource is deleted we create a ClusterDeprovision resource, which launches a pod to run the openshift-install deprovision process until it completes, after which the finalizer is removed and the ClusterDeployment is removed.
+
+There is presently no deprovision implementation for agent based or bare metal clusters but there likely will be in the future.
+
+It stands to reason that deprovision implementations are fully tied to ClusterInstall implementations, and thus deleting the ClusterInstall should trigger the controllers for that ClusterInstall to complete an uninstall/deprovision. *WARNING: Deleting a ClusterInstall will destroy the cluster.*
```

Comment:
Clarified that deleting ClusterInstall will destroy the cluster.

Do we want to assume to add an OwnerRef on the ClusterInstall pointing to ClusterDeployment, so it is automatically garbage collected if the CD is deleted to match current behavior? Typically we do not add owner refs to things that we didn't create for you.

Reply:
It doesn't seem like a ClusterInstall makes sense on its own, and in fact `spec.clusterDeploymentRef` is required. So it makes sense to me to add the owner ref and let it get garbage collected if the CD is gone.